### PR TITLE
New node: Flip

### PR DESCRIPTION
**Pull request by ShouvikGhosh2048**

### CODE REVIEW

**Comment by TrueDoctor** [comment]
@Keavon does it make sense to have all of these options exposed for the flip nodes? I would have probably only added very limited controls and removed options like shear and translation and used a transform node instead.
The Flip node is a strict subset of what the Transform node can do so I would argue that its purpose should be to be a very simple wrapper for the full transform node which is a bit easier to use and as soon as you need more features, you can chain a translate node. But I'm not sure if that is what makes most sense from a users perspective

**Comment by Keavon** [comment]
"Flip X" and "Flip Y" are the two parameters I'd like this node to have.

**Comment by TrueDoctor** [comment]
> "Flip X" and "Flip Y" are the two parameters I'd like this node to have.

And pivot I would assume

**Comment by Keavon** [comment]
> And pivot I would assume

Just within the bounding box I would think.

**Comment by ShouvikGhosh2048** [comment]
The node currently only allows the user to choose `Horizontal` or `Vertical` from a dropdown (I'll change this to `FlipX` and `FlipY`) - the `transform`, `scale`, `rotation`, `skew` and `pivot` inputs are only there to get the current transform, scale, etc. Is there a better way to get these values?

File: editor/src/messages/portfolio/document/node_graph/node_graph_message_handler/document_node_types.rs [hunk 1]
```diff
@@ -1713,6 +1713,24 @@ fn static_nodes() -> Vec<DocumentNodeType> {
...
+			category: "Transform",
+			identifier: NodeImplementation::proto("graphene_core::transform::FlipNode<_, _, _, _, _, _, _>"),
+			inputs: vec![
+				DocumentInputType::value("Data", TaggedValue::VectorData(graphene_core::vector::VectorData::empty()), true),
+				DocumentInputType::value("FlipX", TaggedValue::Bool(false), false),
+				DocumentInputType::value("FlipY", TaggedValue::Bool(false), false),
+				DocumentInputType::value("Translation", TaggedValue::DVec2(DVec2::ZERO), false),
+				DocumentInputType::value("Rotation", TaggedValue::F32(0.), false),
+				DocumentInputType::value("Scale", TaggedValue::DVec2(DVec2::ONE), false),
+				DocumentInputType::value("Skew", TaggedValue::DVec2(DVec2::ZERO), false),
```
Comment (TrueDoctor) [inline editor/src/messages/portfolio/document/node_graph/node_graph_message_handler/document_node_types.rs:1727]:
```suggestion
```

File: node-graph/gcore/src/transform.rs [hunk 2]
```diff
@@ -98,3 +99,32 @@ pub(crate) fn set_transform<Data: TransformMut, TransformInput: Transform>(mut d
...
 }
+
+#[derive(Debug, Clone, Copy)]
+pub struct FlipNode<FlipX, FlipY, Translation, Rotation, Scale, Shear, Pivot> {
+	pub(crate) flip_x: FlipX,
+	pub(crate) flip_y: FlipY,
+	pub(crate) translate: Translation,
+	pub(crate) rotate: Rotation,
+	pub(crate) scale: Scale,
+	pub(crate) shear: Shear,
```
Comment (TrueDoctor) [inline node-graph/gcore/src/transform.rs:110]:
```suggestion
```

File: node-graph/gcore/src/transform.rs [hunk 3]
```diff
@@ -98,3 +99,32 @@ pub(crate) fn set_transform<Data: TransformMut, TransformInput: Transform>(mut d
...
+	let pivot = DAffine2::from_translation(data.local_pivot(pivot));
+	let mut new_scale = scale;
+	if flip_x {
+		new_scale.x *= -1.0;
+	}
+	if flip_y {
+		new_scale.y *= -1.0;
+	}
+
+	let modification = pivot * DAffine2::from_scale_angle_translation(new_scale, rotate as f64, translate) * DAffine2::from_cols_array(&[1., shear.y, shear.x, 1., 0., 0.]) * pivot.inverse();
```
Comment (TrueDoctor) [inline node-graph/gcore/src/transform.rs:125]:
```suggestion
pub(crate) fn flip_vector_data<Data: TransformMut>(mut data: Data, flip_x: bool, flip_y: bool, pivot: DVec2) -> Data {
	let pivot = DAffine2::from_translation(data.local_pivot(pivot));
	let bool_to_scale = |b| if b { -1. } else { 1. }
	let new_scale = DVec2::new(bool_to_scale(flip_x), bool_to_scale(flip_y));

	let modification = pivot * DAffine2::from_scale(new_scalee)  * pivot.inverse();
```

File: node-graph/interpreted-executor/src/node_registry.rs [hunk 4]
```diff
@@ -524,6 +524,9 @@ fn node_registry() -> HashMap<NodeIdentifier, HashMap<NodeIOTypes, NodeConstruct
 		register_node!(graphene_core::transform::SetTransformNode<_>, input: ImageFrame<Color>, params: [ImageFrame<Color>]),
 		register_node!(graphene_core::transform::SetTransformNode<_>, input: VectorData, params: [DAffine2]),
 		register_node!(graphene_core::transform::SetTransformNode<_>, input: ImageFrame<Color>, params: [DAffine2]),
+		register_node!(graphene_core::transform::FlipNode<_, _, _, _, _, _, _>, input: VectorData, params: [bool, bool, DVec2, f32, DVec2, DVec2, DVec2]),
+		register_node!(graphene_core::transform::FlipNode<_, _, _, _, _, _, _>, input: ImageFrame<Color>, params: [bool, bool, DVec2, f32, DVec2, DVec2, DVec2]),
+		register_node!(graphene_core::transform::FlipNode<_, _, _, _, _, _, _>, input: WasmSurfaceHandleFrame, params: [bool, bool, DVec2, f32, DVec2, DVec2, DVec2]),
```
Comment (TrueDoctor) [inline node-graph/interpreted-executor/src/node_registry.rs:529]:
```suggestion
		register_node!(graphene_core::transform::FlipNode<_, _, _>, input: VectorData, params: [bool, bool, DVec2]),
		register_node!(graphene_core::transform::FlipNode<_, _, _>, input: ImageFrame<Color>, params: [bool, bool, DVec2),
		register_node!(graphene_core::transform::FlipNode<_, _, _>, input: WasmSurfaceHandleFrame, params: [bool, bool, DVec2]),
```

**Review by TrueDoctor — CHANGES_REQUESTED** [review]

**Review by TrueDoctor — APPROVED** [review]

**Comment by Keavon** [comment]
Just checking what this is held up on.

**Comment by ShouvikGhosh2048** [comment]
The `update_bounds` drift issue.

**Comment by Keavon** [comment]
@TrueDoctor is that something you'd be able to help with?

**Comment by 0HyperCube** [comment]
!build

**Comment by Keavon** [comment]
Closing this for now since it's been inactive for a while. Feel free to reopen it if it's revived.

**Pull request closed without merging**